### ISSUE TITLE
NUnit is too good

### ISSUE THREAD

**Issue by IanMcRV** [nunit/docs#208]
Dear Charles
Now that parallelism works on the VSAgent in visual studio it's just too good!

can you make it, badder?

**Comment by rprouse** [comment]
@IanMcRV do you have any suggestions on how we can make it badder? We are always looking for input. 😃

**Comment by SamJongenelen** [comment]
Make 'paralizable' inherit from 'Sequential'  and manipulate test run report run times?

**Comment by jnm2** [comment]
@CharliePoole I finally got a chance to flip the switch and use `[assembly: Parallelizable(ParallelScope.Children)]` and wheee! it's night and day! @IanMcRV is not wrong. My life just got so much better.

**Comment by CharliePoole** [comment]
Cool! You should dig into it and see how it works now. 😈

**Comment by jnm2** [comment]
But I only had one single test that wasn't embarrassingly parallel, and now I have no need for anything more granular than the assembly attribute! 😀

You're right, of course... 😁

**Comment by InkUisitor** [comment]
I know that this thread starts by a joke, but I want NUnit to be "badder" to be badass. (Bonjour le France, je pense qu'on dit "worst" et pas "badder")
Everybody uses [assembly: Parallelizable(ParallelScope.Children)] to improve the time of their test. I use the ParallelScope.All for the most of my test. It works correctly. But for some test I don't want to use this Attribute and I don't want to use ParallelScope.None too.
I'm using Parallelization in a Selenium project.
For example, I have classA, classB, classC, classD, classE. 
Methods of classA and classB are not dependent about resources used on others class and methods. I use ParallelScope.All on classA and classB because I want to minimize the time of execution here. 
Methods of classC are too much dependent on the resources, it's too dangerous and destructive for me to use any Parallelization. I use ParallelScope.None on classC.
Methods of classD and classE are dependent inside their own class but methods of classD are not dependent of methods of classE. My problem is here. I don't know what attribute i can use here. The behaviour I want is that classD and classE run in parallel but I don't want to run their own methods in parallel because classD is not destructive on classE but methods of classD are self destructive. Is Children appropriate for this ? I forget to say that the namespace of my class are not everytime the same, can it be a problem ?

**Comment by jnm2** [comment]
Would https://github.com/nunit/nunit/issues/2389 provide the flexibility you want?

**Comment by InkUisitor** [comment]
What I wanted to do is something like that :

Loop | Agent1 | Agent2 | Agent 3 | Agent4
--------|-------|----------|--------|------
1 | A | D | B | A
2 | A | A | A | B
3 | A | D | E | A
4 | D | B | B | E
5 | D | E | |
6 | D | | |
7 | C | | |
8 | C | | |

But actually the best I can do is :

Loop | Agent1 | Agent2 | Agent 3 | Agent4
--------|-------|----------|--------|------
1 | A | B | B | A
2 | A | A | A | B
3 | A | B | B | A
5 | D | E | |
4 | E | D | |
5 | D | E | |
6 | D | | |
7 | C | | |
8 | C | | |

It seems that your proposition is not what I look at for. It's interesting, but I don't feel the need actually.
I feel like i want :
classA, classB --> do a ParallelScope.All
classC --> do a ParallelScope.None
classD, classE --> do a ParallelScope.All but use only one agent by class. Maybe a "ParallelScope.AllUnique"
Or maybe I'm missing something.
The concept is maybe weird and you probably assume that this kind of scenario will not be covered. An other problem will be that the organisation of tests with this flag can be difficult. Will we optimise the order of the test (but the test can become flaw against a random organisation ?)... I don't know...

I don't feel any priority about that, but if this feature exists, please share me the tips ! 🕺 

Do you want me to open another issue for that ?

**Comment by InkUisitor** [comment]
Oops, don't see I was on nunit/doc 🤦‍♂️

**Comment by jnm2** [comment]
@InkUisitor If I understand right, you want:

 - A and B to run concurrently with A, B, D, and E (but not C)
 - D to run concurrently with A, B and E but not other Ds (and not C)
 - E to run concurrently with A, B and D but not other Es (and not C)
 - C to run concurrently with nothing, including itself

You can already accomplish the A, B and C interactions with NUnit today. 
To achieve the D and E interactions, nunit/nunit#2389 would give you that ability by marking the D fixture with `[NotParallelizable("D group (for example")]` and marking the E fixture with  `[NotParallelizable("E group (for example")]`.

**Comment by CharliePoole** [comment]
@InkUisitor I marked this as a question. Although we do take questions here, especially about the documentation, you will usually get a better response on our Google discussion group `nunit-discuss` It's monitored by more people.

We have a number of outstanding bugs around parallelism. I'm going to describe what is __supposed__ to work, without 100% guarantee that it isn't broken. If you try it and find it doesn't work as described, an issue on the nunit framework repository (https://github.com/nunit/nunit) would be the thing to do if there isn't one already there.

One other small thing - so we can talk the same language. The term `agent` as used by NUnit relates to running tests in a separate process. It has nothing to do with the `ParallelizableAttribute`.

In __theory__ this should work... at least as I understand what you are trying to do...

```
[Parallelizable(ParallelScope.All)]
public class ClassA{...}

[Parallelizable(ParallelScope.All)]
public class ClassB{...}

[NonParallelizable]
public class ClassC{...}

namespace SomeNamespace
{
    [SetUpFixture, NonParallelizable]
    public class DummySetUpFixture{...}

    [Parallelizable(ParallelScope.All)]
    public class ClassA{...}

    [Parallelizable(ParallelScope.All)]
    public class ClassA{...
}
```

Let me repeat and reinforce: this is not only __not guaranteed__ to work but I pretty much don't expect it to work based on the code. But it's supposed to work and I'm in the middle of a big PR that is designed to make it work.

However, I'd love to see you give it a try and report the results!

**Comment by CharliePoole** [comment]
BTW issue nunit/nunit#2389 would also resolve this for you, but it's aimed at much more complex, non-hierarchical situations than yours. The basic hierarchical structure of tests is supposed to take care of what you are trying to accomplish.

Another bit of advice... using `ParallelScope.Children` at the assembly level is only a good idea if your application is one where almost all tests can run in parallel. I'm a little surprised that you say "Everybody" uses it and I hope that's not true. Writing tests that can all run in parallel is a bit tricky and I don't think most people do it.

**Comment by InkUisitor** [comment]
@jnm2 That's exactly what I'm looking for.
@CharliePoole I thought the same things than you about agents. Today, I'm running a Selenium campain with 16 nunit agents. I tried to represent the running of my setup with 4 agents only and a cell was representing a method. But representing the behaviour of that with a table is maybe not the good way to do it. I think that the issue #165 you opened will clearly help me about "my problem". Today, I can do it in a less optimise way but it's okay for the moment. I will follow your ticket to see modifications about that behaviour.
To finish, I often see the parallelscope just defined in the assembly when you look at sites that talks about NUnit. When everybody talks about NUnit, you feels like it's a magic tool that can parallelise all your scenario without explaining you about the different ParallelScope. And the facts shows that it's more complicated than "JUST DO IT !!!".
I will look at the google group to get more informations on my daily questions.

**Comment by jnm2** [comment]
> BTW issue nunit/nunit#2389 would also resolve this for you, but it's aimed at much more complex, non-hierarchical situations than yours. The basic hierarchical structure of tests is supposed to take care of what you are trying to accomplish.

@CharliePoole To me, parallelism is all about opt-out, specifically in terms of `why` do I have to opt out for a given test? The `why` goes in the `[NotParallelizable(MyScopeStrings.BecauseItUsesGlobalStateX)]`

I can't wrap my head around how I would set this up using only the hierarchical structure of tests:

>  - D to run concurrently with A, B and E but not other Ds (and not C)
>  - E to run concurrently with A, B and D but not other Es (and not C)

Maybe once you explain it I'll feel silly though. 😃

**Comment by InkUisitor** [comment]
@jnm2 & @CharliePoole 
I'm a rookie about testing, and I'm learning everyday. I remember I said that Google made sh*t with their integration of their Chrome driver with Selenium. I never thought that : "Hey man, you use NUnit like a sh*t and that's exactly why you have errors in your test campain like that".
I have set my TestFixtures in order to answer a functional vision and not a technical vision with Selenium. It's way more easier to read the scenario like that. I don't design technical test the same way because I usually want a transversal vision.
I think that my vision with Selenium project gives me problems for the parallelisation. But today, I prefer to have problems with the parallelisation than have ones about reading the main functionnality of my test.
My classC correspond to make different operations on the licensing of my soft.
My classD correspond to make different operations on the SMTP management of the server and E is about make operations on my legal informations.
Maybe this kind of vision is special for standard testing, but I can often see problems like that these days because of Selenium.
When I saw [NotParallelizable(MyScopeStrings.BecauseItUsesGlobalStateX)] I felt confortable to explain that my TestFixture will not run its methods in parallel because it UsesGlobalStateX. I don't know how I would write this with a ParallelScope.MagicAttribute. Or maybe like a token : ParallelScope.ParallelToken("UseGlobalStateX", canExecuteWith : {ParallelScope.All, "UseGlobalStateY"})

**Comment by CharliePoole** [comment]
@jnm2 Cleaning out old notices and came across this...

"To me, parallelism is all about opt-out..."

Just so you know, this was discussed at the time we introduced parallelism within the framework and we specifically decided as a community that it should be about opt-in, not opt-out.

There's a lot to be said for your point of view in the ideal situation: that is, if we had been designing a parallelism mechanism for test execution without regard to existing tests. It's actually the point of view that I plan to take in my future work on test frameworks. It's just (at least IMO) off topic for discussions of how to make incremental improvements to NUnit as it stands.

In retrospect, I halfway wish we had not allowed the user to specify parallelism by default on the assembly, since it appears to mislead them into thinking that NUnit will do something "extra" to make things that won't normally work in parallel work!

**Comment by SeanKilleen** [comment]
🧹 Housekeeping: Closing this as an older issue with no discernable actionable docs follow-up.
